File: web/src/lib/types/omni-tool-result.ts
/// Typed envelopes for structured tool_result content blocks.
///
/// Mirrors `services/ai/tools/omni_tool_result.py`. When a tool surfaces a
/// UI-driven prompt (e.g. "user must complete OAuth") rather than a normal
/// action result, the AI service encodes a typed envelope inside the
/// tool_result's text content. The frontend parses it here and dispatches on
/// `omni_kind` to render the right card.

export const OmniToolResultKind = {
    OauthRequired: 'oauth_required',
} as const

export type OmniToolResultKind = (typeof OmniToolResultKind)[keyof typeof OmniToolResultKind]

export type OAuthRequiredPayload = {
    source_id: string
    source_type: string
    provider: string
    /** Relative URL on the omni-web host, e.g. `/api/oauth/start?source_id=...`. */
    oauth_start_url: string
}

export type OmniToolResultEnvelope = {
    omni_kind: typeof OmniToolResultKind.OauthRequired
    payload: OAuthRequiredPayload
}

/**
 * Best-effort parse of a tool_result text block as an Omni envelope. Returns
 * null if the text isn't a recognized envelope (in which case the caller
 * should render the text as-is).
 */
export function tryParseOmniEnvelope(text: string): OmniToolResultEnvelope | null {
    if (!text || text[0] !== '{') return null
    let obj: unknown
    try {
        obj = JSON.parse(text)
    } catch {
        return null
    }
    if (!obj || typeof obj !== 'object') return null
    const candidate = obj as Record<string, unknown>
    const kind = candidate.omni_kind
    const payload = candidate.payload
    if (typeof kind !== 'string' || !payload || typeof payload !== 'object') {
        return null
    }
    if (kind === OmniToolResultKind.OauthRequired) {
        const p = payload as Record<string, unknown>
        if (
            typeof p.source_id !== 'string' ||
            typeof p.source_type !== 'string' ||
            typeof p.provider !== 'string' ||
            typeof p.oauth_start_url !== 'string'
        ) {
            return null
        }
        return {
            omni_kind: OmniToolResultKind.OauthRequired,
            payload: {
                source_id: p.source_id,
                source_type: p.source_type,
                provider: p.provider,
                oauth_start_url: p.oauth_start_url,
            },
        }
    }
    return null
}

File: services/ai/tools/omni_tool_result.py
"""Typed envelopes for structured tool_result content blocks.

When a tool needs to surface a UI-driven prompt (e.g. "user must complete OAuth")
rather than a normal action result, we encode a typed envelope inside the
tool_result's text content. The frontend parses this envelope and renders the
appropriate prompt; if the LLM ever sees the placeholder it sees machine-readable
JSON rather than misleading prose.

Anthropic's tool_result content array only accepts `text` and `image` block
types — JSON-in-text is the cleanest portable shape for structured signals.
"""

from __future__ import annotations

import json
import logging
from dataclasses import asdict, dataclass
from enum import Enum

from anthropic.types import TextBlockParam

logger = logging.getLogger(__name__)


class OmniToolResultKind(str, Enum):
    OAUTH_REQUIRED = "oauth_required"
    APPROVAL_REQUIRED = "approval_required"


@dataclass
class ApprovalRequiredPayload:
    approval_id: str
    tool_name: str
    tool_input: dict[str, object]
    tool_call_id: str
    source_id: str | None = None
    source_type: str | None = None


@dataclass
class OAuthRequiredPayload:
    """Surfaced when a connector action returns 412 needs_user_auth.

    `oauth_start_url` is relative (e.g. `/api/oauth/start?source_id=...`); the
    web layer prefixes the host before opening the popup.
    """

    source_id: str
    source_type: str
    provider: str
    oauth_start_url: str


def _encode_envelope(kind: OmniToolResultKind, payload: object) -> TextBlockParam:
    return TextBlockParam(
        type="text",
        text=json.dumps(
            {
                "omni_kind": kind.value,
                "payload": asdict(payload),
            }
        ),
    )


def encode_approval_required(payload: ApprovalRequiredPayload) -> TextBlockParam:
    """Wrap an ApprovalRequiredPayload as a tool_result text content block."""
    return _encode_envelope(OmniToolResultKind.APPROVAL_REQUIRED, payload)


def encode_oauth_required(payload: OAuthRequiredPayload) -> TextBlockParam:
    """Wrap an OAuthRequiredPayload as a tool_result text content block."""
    return _encode_envelope(OmniToolResultKind.OAUTH_REQUIRED, payload)
